store/store.go

```diff
@@ -499,6 +504,51 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
+// try advance contiguous head based on already written headers.
+func (s *Store[H]) advanceContiguousHead(ctx context.Context) {
+	currHead := s.contiguousHead.Load()
+	if currHead == nil {
+		return
+	}
+	currHeight := (*currHead).Height()
+	prevHeight := currHeight
+
+	// TODO(cristaloleg): benchmark this timeout or make it dynamic.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	var newHead H
+	for {
+		h, err := s.getByHeight(ctx, currHeight+1)
+		if err != nil {
+			break
```

Comment:
The downside I realized right now about the single loop approach I suggested earlier. Basically every write now involves a read. The curreHeight+1 is gonna check if we have header or not on every advancement try, which might be expensive. The good thing is that we do that in flushLoop so its less of an issue, but we should keep that in mind anyway.

Reply:
Partially yes. However, `getByHeight` uses caches (`pending`, `heightIndex`) heavily, in most cases we will not hit the disk at all (I'm literally repeating your words now).

Reply:
The last `getByHeight` will always hit disk, because the last head is not in any of caches

Reply:
anything to do here?

Reply:
This depends on where we put head advancing in #241. If we do it in `Append`, the additional read is going to be detrimental to `Append'`s performance, when its already does a bunch of IO. If we don't find a way to put it into `flushLoop` preserving synchrony, we need to optimize it then.